gtsam/basis/Basis.h

```diff
@@ -0,0 +1,381 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file Basis.h
+ * @brief Compute an interpolating basis
+ * @author Varun Agrawal
+ * @date July 4, 2020
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/OptionalJacobian.h>
+#include <gtsam/basis/ParameterMatrix.h>
+
+#include <gtsam/3rdparty/Eigen/unsupported/Eigen/KroneckerProduct>
+#include <iostream>
+
+/**
+ *  Concept of a Basis:
+ *    - type Weights, Parameters
+ *    - CalculateWeights(size_t N, double a=default, double b=default)
+ */
+
+namespace gtsam {
+
+/// CRTP Base class for function bases
+template <typename DERIVED>
+class Basis {
+ public:
+  /// Call weights for all x in vector, returns M*N matrix where M is the size
+  /// of the vector X.
+  static Matrix WeightMatrix(size_t N, const Vector& X) {
+    Matrix W(X.size(), N);
+    for (int i = 0; i < X.size(); i++)
+      W.row(i) = DERIVED::CalculateWeights(N, X(i));
+    return W;
+  }
+
+  /// Call weights for all x in vector, with interval [a,b]
+  /// Returns M*N matrix where M is the size of the vector X.
+  static Matrix WeightMatrix(size_t N, const Vector& X, double a, double b) {
+    Matrix W(X.size(), N);
+    for (int i = 0; i < X.size(); i++)
+      W.row(i) = DERIVED::CalculateWeights(N, X(i), a, b);
+    return W;
+  }
+
+  /**
+   * EvaluationFunctor at a given x, applied to Parameters.
+   * This functor is used to evaluate a parameterized function at a given scalar
+   * value x. When given a specific N*1 vector of Parameters, returns the scalar
+   * value of the function at x, possibly with Jacobian wrpt the parameters.
+   */
+  class EvaluationFunctor {
+   protected:
+    typename DERIVED::Weights weights_;
```

Comment:
Again, looking at all derived classes, they just say

```
using Weights = Eigen::Matrix<double, /*1xN*/ 1, -1>;
```

i.e., Vector. So just define

```
using Weights = Vector;
```

front of Basis, and remove all mentions of DERIVED::Weights, as well as all typedefs in derived classes.

Reply:
What if we use a parameterization that doesn't use a `Vector`? Code for the now and not the future?